chain/network/src/concurrency/scope/mod.rs

```diff
@@ -0,0 +1,459 @@
+        match self.0.upgrade() {
+            Some(strong) => JoinHandle(
+                Inner::spawn(strong, unsafe { to_static(f.boxed()) }),
+                std::marker::PhantomData,
+            ),
+            // Upgrade may fail only if all the "main" tasks have already completed
+            // so the caller is a "background" task. In that case we fall back
+            // to spawning a "background" task instead. It is ok, since the distinction
+            // between main task and background task disappears, once the scope is canceled.
+            None => self.spawn_bg(f),
+        }
+    }
+
+    /// Spawns a "background" task in the scope.
+    /// It behaves just like a single-task Service, but
+    /// has the same lifetime as the Scope, so it can spawn
+    /// more tasks in the scope. It is not a "main" task, so
+    /// it doesn't prevent scope cancelation.
+    pub fn spawn_bg<T: 'static + Send>(
+        &self,
+        f: impl 'env + Send + Future<Output = Result<T, E>>,
+    ) -> JoinHandle<'env, T> {
+        JoinHandle(
+            Inner::spawn(self.1.upgrade().unwrap(), unsafe { to_static(f.boxed()) }),
+            std::marker::PhantomData,
+        )
+    }
+
+    /// Spawns a service.
+    ///
+    /// Returns a handle to the service, which allows spawning new tasks within the service.
+    pub fn new_service(&self) -> Service<E> {
+        Inner::new_service(self.0.upgrade().unwrap().0.clone())
+    }
+}
+
+/// must_complete wraps a future, so that it aborts if it is dropped before completion.
+///
+/// Possibility that a future can be dropped/aborted at every await makes the control flow unnecessarily complicated.
+/// In fact, only few basic futures (like io primitives) actually need to be abortable, so
+/// that they can be put together into a tokio::select block. All the higher level logic
+/// would greatly benefit (in terms of readability and bug-resistance) from being non-abortable.
+/// Rust doesn't support linear types as of now, so best we can do is a runtime check.
+fn must_complete<Fut: Future>(fut: Fut) -> impl Future<Output = Fut::Output> {
+    let guard = MustCompleteGuard;
+    async move {
+        let res = fut.await;
+        let _ = std::mem::ManuallyDrop::new(guard);
+        res
+    }
+}
+
+struct MustCompleteGuard;
+
+impl Drop for MustCompleteGuard {
+    fn drop(&mut self) {
+        // We always abort here, no matter if compiled with panic=abort or panic=unwind.
+        eprintln!("dropped a non-abortable future before completion");
+        eprintln!("backtrace:\n{}", std::backtrace::Backtrace::force_capture());
+        std::process::abort();
+    }
+}
+
+/// Should be used only via run! macro.
+#[doc(hidden)]
+pub mod internal {
+    use super::*;
+
+    pub fn new_scope<'env, E: 'static>() -> Scope<'env, E> {
+        Scope(Weak::new(), Weak::new(), std::marker::PhantomData)
+    }
+
+    pub async fn run<'env, E, T, F, Fut>(scope: &'env mut Scope<'env, E>, f: F) -> Result<T, E>
```

Comment:
I don’t have full clarity in my mind yet, but I’d expect this should probably be marked `unsafe`? Even though it’s `doc(hidden)` it’d still lower the chances it gets used wrong.

Also… I don’t see how this prevents (tasks spawned from) `f` from executing for longer than `'env`, which is essential for the `to_static` calls above to be correct.

To the best of my current understanding, rustc could (and probably will given the definition of `new_scope`) resolve `'env` to be the empty lifetime, and then eg. if `f` is `'a` then the `to_static` call will expand its lifetime, and it could result in a dangling pointer dereference. But I may be missing something?

If so, maybe a `// SAFETY` comment on the `unsafe` blocks calling `to_static` would help understand what’s going on that actually makes this safe?

Reply:
Scope is invariant in 'env. Since Scope is only constructible via run! macro, 'env is also the lifetime of the Scope object. Scope object is not dropped until the tasks spawned within run() execution complete. So it is more like Scope is being kept alive at least until the tasks complete.

Reply:
Hmmmm yes ok I think I see how that works. Not 100% sure this is correct, but it’s most likely better than actix already, so… 🤷 😅